### PR TITLE
[WIP] dynamic lean

### PR DESCRIPTION
**Pull request by johanveenpere**
Added dynamic leaning. Currently it's only missing prediction code, has problems detecting some collisions(mainly props) and crouching while leaning may jitter sometimes. https://youtu.be/3Fw0MJW8Lw8

### CODE REVIEW

File: mp/src/game/shared/neo/neo_predicted_viewmodel.cpp [hunk 1]
```diff
@@ -164,273 +175,119 @@ bool IsThereRoomForLeanSlide(CNEO_Player *player,
...
+float CNEOPredictedViewModel::calculateLeanAngle(float freeRoom, CNEO_Player *player){
+	Vector	hipBonePosition;
+	QAngle hipBoneAngle;
+	//it would be nicer if we could take the position of hip bone and use that to calculate hipToHeadHeight...
+	/*GetBonePosition(0, hipBonePosition, hipBoneAngle);
+	debugoverlay->AddTextOverlay(hipBonePosition, 0.001, "hip");
+	const float hipToHeadHeight = player->GetViewOffset().z - (hipBonePosition.z - player->GetAbsOrigin().z);*/
+	float hipToHeadHeight = 41;
+	if (player->m_nButtons & IN_DUCK){
+		hipToHeadHeight = 24;
```
Comment (Rainyan) [inline mp/src/game/shared/neo/neo_predicted_viewmodel.cpp]: Changing _hipToHeadHeight_ based on the IN_DUCK state seems to change the final lean angle, meaning you'll lean further away whilst ducking. This feels especially weird if you hold the lean and simultaneously duck-unduck, as you'll be switching between two different max. lean angles.

Also, the head<-->hip distance shouldn't really change for a humanoid regardless of the pose, assuming we're rolling purely around the hip (unless I'm misunderstanding what it's representing). Having this variable be a single constant regardless of duck state might make more sense?

**Comment by Rainyan** [comment]
Cool stuff! There seems to be some jitter in X and Y coords of the eyepos (_cl_showpos 1_) when adding _net_fakelag_ that needs some looking into, but the collision resolution seems much better overall.

File: mp/src/game/shared/neo/neo_predicted_viewmodel.cpp (hunk 1 above)
Comment (johanveenpere) [reply mp/src/game/shared/neo/neo_predicted_viewmodel.cpp]: Technically the distance from bone to viewpoint changes when crouching a little but having it as a constant is an acceptable simplification. I have already committed the changes.

File: mp/src/game/shared/neo/neo_predicted_viewmodel.cpp [hunk 2]
```diff
@@ -129,308 +127,125 @@ inline void CNEOPredictedViewModel::DrawRenderToTextureDebugInfo(IClientRenderab
...
-		m_angNextViewAngles.z = targetAng.z;
-	}
-
-#ifdef CLIENT_DLL
-	if (!prediction->InPrediction() && prediction->IsFirstTimePredicted())
-#endif
-	{
-		// This provides us with lag compensated out values.
-		CalcViewModelLag(m_vecNextViewOffset, m_angNextViewAngles, startAng);
+	else{
```
Comment (Rainyan) [inline mp/src/game/shared/neo/neo_predicted_viewmodel.cpp]: This else condition triggers even when (dY == 0), which seems to be the cause of the jitter when emulating roundtrip lag with the net_fakelag cvar.

Checking if dY == 0 before running the logic would seem to solve it.
```
if (dY != 0) // when we enter this block, the if--else implies larger or lesser, never equal
	{
		if (dY > 0){
			Ycurrent += Ymoved;
			if (Ycurrent >= Yfinal - 0.05f){
				Ycurrent = Yfinal;
			}
		}
		else{
			Ycurrent -= Ymoved;
			if (Ycurrent <= Yfinal + 0.05f){
				Ycurrent = Yfinal;
			}
		}
	}
```

**Pull request merged**